src/connector/src/sink/file_sink/gcs.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use std::collections::{BTreeMap, HashMap};
+
+use anyhow::anyhow;
+use opendal::layers::{LoggingLayer, RetryLayer};
+use opendal::services::Gcs;
+use opendal::Operator;
+use serde::Deserialize;
+use serde_with::serde_as;
+use with_options::WithOptions;
+
+use super::opendal_sink::FileSink;
+use crate::sink::file_sink::opendal_sink::OpendalSinkBackend;
+use crate::sink::{Result, SinkError, SINK_TYPE_APPEND_ONLY, SINK_TYPE_OPTION, SINK_TYPE_UPSERT};
+
+#[derive(Deserialize, Debug, Clone, WithOptions)]
+pub struct GcsCommon {
+    #[serde(rename = "gcs.bucket_name")]
+    pub bucket_name: String,
+
+    /// The base64 encoded credential key. If not set, ADC will be used.
+    #[serde(rename = "gcs.credential")]
+    pub credential: Option<String>,
+
+    /// If credential/ADC is not set. The service account can be used to provide the credential info.
+    #[serde(rename = "gcs.service_account", default)]
+    pub service_account: Option<String>,
+
+    /// The directory where the sink file is located
+    #[serde(rename = "gcs.path", default)]
+    pub path: String,
+
+    #[serde(flatten)]
+    pub unknown_fields: HashMap<String, String>,
+}
+
+#[serde_as]
+#[derive(Clone, Debug, Deserialize, WithOptions)]
+pub struct GcsConfig {
+    #[serde(flatten)]
+    pub common: GcsCommon,
+
+    pub r#type: String, // accept "append-only" or "upsert"
+}
+
+pub const GCS_SINK: &str = "gcs";
+
+impl<S: OpendalSinkBackend> FileSink<S> {
+    pub fn new_gcs_sink(config: GcsConfig) -> Result<Operator> {
+        // Create gcs builder.
+        let mut builder = Gcs::default();
+
+        builder.bucket(&config.common.bucket_name);
+
+        // if credential env is set, use it. Otherwise, ADC will be used.
+        if let Some(cred) = config.common.credential {
+            builder.credential(&cred);
+        } else {
+            let cred = std::env::var("GOOGLE_APPLICATION_CREDENTIALS");
```

Comment:
For RW running on GCP, will `GOOGLE_APPLICATION_CREDENTIALS` be set automatically? If yes, gcs sink may mistakenly write to the RW S3. Should we make `credential` required instead of optional?

Reply:
I think you want to completely isolate the RW gcs and sink's object store, right? SGTM.

Reply:
Yes. We isolate RW S3 and sink's S3 if the object store is S3 already so it is better to keep it consistent for other object store types.

Reply:
I think we need to make `service_account` required, not `credential` required.
- For cloud user: because a pod can only have one workload identity, unless the workload identity is given rw permissions to access the sink gcs, the sink gcs can only use another service account through static credential.
- For open source user: it is entirely possible that the same service account has access to both the gcs used by rw and the gcs of the downstream sink into, so the same credential can be used.

If we make `service account` required, user will write to the same bucket only when the service account and rw gcs are consistent. This behavior is determined by the user.

And `credential` field only provides multiple auth options.

Reply:
What configs should user provide if he/she wants to use `service_account`?

Reply:
After some offline discussion, let's make both `service_account` and `credential` required for safety.